File: app/Services/SubscriptionService.php
<?php

namespace App\Services;

use App\Models\User;
use App\Models\Subscription;
use Carbon\Carbon;
use ValueError;

class SubscriptionService
{
    const PLAN_PRICES = [
        'premium_simple' => 10.0,
        'premium_deluxe' => 30.0,
    ];

    const PLAN_DURATION_DAYS = 365; // 1 an

    /**
     * Get max products for a subscription plan
     */
    public static function getMaxProductsForPlan(string $plan): int
    {
        return match ($plan) {
            'FREE' => 1,
            'PREMIUM_SIMPLE' => 1,
            'PREMIUM_DELUXE' => 999999, // Unlimited
            default => 1,
        };
    }

    /**
     * Check if user can add a product
     */
    public static function canAddProduct(User $user): bool
    {
        // Check if subscription is active
        if ($user->subscription_tier !== 'FREE') {
            if (!$user->subscription_end_date || $user->subscription_end_date < now()) {
                // Subscription expired, revert to free
                $user->subscription_tier = 'FREE';
                $user->subscription_end_date = null;
                $user->save();
            }
        }

        $maxProducts = self::getMaxProductsForPlan($user->subscription_tier);
        $currentCount = $user->products()->where('is_active', true)->count();

        return $currentCount < $maxProducts;
    }

    /**
     * Check if user can add an alert
     */
    public static function canAddAlert(User $user): bool
    {
        // For now, alerts are unlimited for all plans
        // Can be adjusted based on business requirements
        return true;
    }

    /**
     * Create a new subscription
     */
    public static function createSubscription(
        User $user,
        string $planName,
        ?string $paymentReference = null
    ): Subscription {
        if (!isset(self::PLAN_PRICES[$planName])) {
            throw new ValueError("Invalid plan: {$planName}");
        }

        // Create subscription
        $subscription = Subscription::create([
            'user_id' => $user->id,
            'plan' => $planName,
            'status' => 'active',
            'amount' => self::PLAN_PRICES[$planName],
            'currency' => 'EUR',
            'starts_at' => now(),
            'expires_at' => now()->addDays(self::PLAN_DURATION_DAYS),
            'payment_reference' => $paymentReference,
        ]);

        // Update user
        if ($planName === 'premium_simple') {
            $user->subscription_tier = 'PREMIUM_SIMPLE';
        } elseif ($planName === 'premium_deluxe') {
            $user->subscription_tier = 'PREMIUM_DELUXE';
        }

        $user->subscription_expires_at = $subscription->expires_at;
        $user->subscription_start_date = $subscription->starts_at;
        $user->save();

        return $subscription;
    }

    /**
     * Get allowed alert channels for user
     */
    public static function getAllowedAlertChannels(User $user): array
    {
        if ($user->subscription_tier === 'FREE') {
            return ['email'];
        } else {
            return ['email', 'whatsapp', 'push'];
        }
    }

    /**
     * Get max alerts for plan
     */
    public static function getMaxAlertsForPlan(string $plan): int
    {
        // For now, unlimited for all plans
        return 999999;
    }
}

File: services/subscription_service.py
from models.user import User
from models.subscription import Subscription
from datetime import datetime, timedelta
from extensions import db

class SubscriptionService:
    """Service for managing subscriptions"""
    
    PLAN_PRICES = {
        'premium_simple': 10.0,
        'premium_deluxe': 30.0,
    }
    
    PLAN_DURATION_DAYS = 365  # 1 year
    
    @staticmethod
    def get_max_products_for_plan(plan: str) -> int:
        """Get max products for a subscription plan"""
        mapping = {
            'FREE': 1,
            'PREMIUM_SIMPLE': 1,
            'PREMIUM_DELUXE': 999999,  # Unlimited
        }
        return mapping.get(plan, 1)
    
    @staticmethod
    def can_add_product(user: User) -> bool:
        """Check if user can add a product"""
        # Check if subscription is active
        if user.subscription_tier != 'FREE':
            if not user.subscription_end_date or user.subscription_end_date < datetime.utcnow():
                # Subscription expired, revert to free
                user.subscription_tier = 'FREE'
                user.subscription_end_date = None
                db.session.commit()
        
        max_products = SubscriptionService.get_max_products_for_plan(user.subscription_tier)
        from models.product import Product
        current_count = Product.query.filter_by(user_id=user.id, is_active=True).count()
        
        return current_count < max_products
    
    @staticmethod
    def can_add_alert(user: User) -> bool:
        """Check if user can add an alert"""
        # For now, alerts are unlimited for all plans
        return True
    
    @staticmethod
    def create_subscription(user: User, plan_name: str, payment_reference: str = None) -> Subscription:
        """Create a new subscription"""
        if plan_name not in SubscriptionService.PLAN_PRICES:
            raise ValueError(f"Invalid plan: {plan_name}")
        
        # Create subscription
        subscription = Subscription(
            user_id=user.id,
            plan=plan_name,
            status='active',
            amount=SubscriptionService.PLAN_PRICES[plan_name],
            currency='EUR',
            starts_at=datetime.utcnow(),
            expires_at=datetime.utcnow() + timedelta(days=SubscriptionService.PLAN_DURATION_DAYS),
            payment_reference=payment_reference
        )
        db.session.add(subscription)
        
        # Update user
        if plan_name == 'premium_simple':
            user.subscription_tier = 'PREMIUM_SIMPLE'
        elif plan_name == 'premium_deluxe':
            user.subscription_tier = 'PREMIUM_DELUXE'
        
        user.subscription_start_date = subscription.starts_at
        user.subscription_end_date = subscription.expires_at
        db.session.commit()
        
        return subscription
    
    @staticmethod
    def get_allowed_alert_channels(user: User) -> list:
        """Get allowed alert channels for user"""
        if user.subscription_tier == 'FREE':
            return ['email']
        else:
            return ['email', 'whatsapp', 'push']
    
    @staticmethod
    def get_max_alerts_for_plan(plan: str) -> int:
        """Get max alerts for plan"""
        # For now, unlimited for all plans
        return 999999

